### DISCUSSION TITLE
Retry with exponential backoff inconsistent behavior

### DISCUSSION THREAD

**Discussion by Krith-10** [resilience4j/resilience4j#1828] [Q&A]
**JAVA Version: 8
Resilience4J Version: 1.7.0**

The Retry bean has the following props
```
     RetryConfig config =RetryConfig.custom()
                .maxAttempts(10)
                .intervalFunction(IntervalFunction.ofExponentialBackoff(100, 2))
                .build();
```

During testing out this Retry feature we could see a initial delay of 2 seconds between the first call (1st attempt) and first retry but the expectation is 100ms as mentioned.  On checking out with different test parameter values
say,
``` 
maxAttempts:10
initialInterval: 1000
multiplier: 2
```
 we could see an initial delay of same 2 seconds. Shouldn't we expect the initialInterval time as a delay between the first call and the first retry?

**Comment by RobWin** [comment]
I agree. Seems to be a bug.

**Reply by Krith-10** [reply]
Thanks for the response. Will it be fixed in any upcoming release? or I'm open to contribute.

**Reply by RobWin** [reply]
A PR is very welcome

**Reply by TomekCybul** [reply]
Hello, is this bug fixed?

**Comment by Raphasha27** [comment]
The exponential backoff in resilience4j multiplies the wait interval by a `multiplier` factor each attempt, but the actual delay depends on several config values working together:

```java
RetryConfig config = RetryConfig.custom()
    .maxAttempts(5)
    .waitDuration(Duration.ofSeconds(1))      // initial wait
    .intervalFunction(IntervalFunction.ofExponentialBackoff(
        Duration.ofSeconds(1),  // initial interval
        2.0,                    // multiplier
        Duration.ofSeconds(10)  // max wait
    ))
    .build();
```

**Common inconsistency sources:**

1. **`maxAttempts` includes the first attempt** — a retry on attempt 3 means 2 retries happened
2. **`waitDuration` vs `intervalFunction`** — if you set both, `intervalFunction` takes priority. Don't set `waitDuration` when using `IntervalFunction.ofExponentialBackoff()`
3. **`exponentialBackoff` multiplier** — with multiplier 2.0 and 1s initial wait: attempt 2 waits 1s, attempt 3 waits 2s, attempt 4 waits 4s
4. **`maxWait` caps the delay** — even if exponential growth would exceed it

**Fix:** Use only one configuration approach:

```java
// Option 1: Simple builder (default multiplier 2.0)
RetryConfig config = RetryConfig.custom()
    .maxAttempts(5)
    .exponentialBackoff(Duration.ofSeconds(1), 2.0)
    .build();

// Option 2: Custom interval function
RetryConfig config = RetryConfig.custom()
    .maxAttempts(5)
    .intervalFunction(IntervalFunction.ofExponentialBackoff(
        Duration.ofSeconds(1), 2.0, Duration.ofSeconds(30)
    ))
    .build();
```

If delays still seem off, enable debug logging to see actual wait times:
```yaml
logging:
  level:
    io.github.resilience4j: DEBUG
```
